crucible-mir/src/Mir/TransTy.hs

```diff
@@ -669,41 +764,77 @@ structInfo adt i = do
         Just fld -> return $ fld ^. M.fty
         Nothing -> mirFail errFieldIndex
 
-    Some ctx <- variantFieldsM var
-    Some idx <- case Ctx.intIndex (fromIntegral i) (Ctx.size ctx) of
-        Just x -> return x
-        Nothing -> mirFail errFieldIndex
-    let tpr' = ctx Ctx.! idx
-    Some tpr <- tyToReprM fldTy
-
     col <- use $ cs . collection
-    kind <- checkFieldKind (not $ canInitialize col fldTy) tpr tpr' $
-        "field " ++ show i ++ " of struct " ++ show (adt ^. M.adtname)
+    case adtSizedness col adt of
+      Sized -> do
+        Some ctx <- variantFieldsM var
+        Some idx <- case Ctx.intIndex (fromIntegral i) (Ctx.size ctx) of
+            Just x -> return x
+            Nothing -> mirFail errFieldIndex
+        let tpr' = ctx Ctx.! idx
+        Some tpr <- tyToReprM fldTy
+
+        kind <- checkFieldKind (not $ canInitialize col fldTy) tpr tpr' $
+            "field " ++ show i ++ " of struct " ++ show (adt ^. M.adtname)
+
+        return $ SizedStruct ctx idx kind
+
+      -- We want to avoid attempting to compute the `CtxRepr` for this struct,
+      -- which may involve computing the `TypeRepr` of its unsized field, which
+      -- does (for `TyDynamic`) or should (for `TyStr` and `TySlice`) cause an
+      -- error.
+      Unsized -> case tySizedness col fldTy of
+        Sized -> do
+          Some (FieldRepr fieldKind) <- pure $ tyToFieldRepr col fldTy
+          pure $ SizedField fieldKind
+        Unsized
+          | i /= length (var ^. M.vfields) - 1 ->
+            mirFail "encountered unsized field in non-last position"
+        Unsized ->
+          case fldTy of
+            M.TySlice innerTy -> do
+              Some innerRepr <- tyToReprM innerTy
+              pure $ UnsizedSliceField innerRepr
+            M.TyStr -> do
+              Some innerRepr <- tyToReprM (M.TyUint M.B8)
+              pure $ UnsizedSliceField innerRepr
+            _ -> pure UnsizedNonSliceField
```

Comment:
Style nit pick: use an `otherwise` guard here so that it's more obvious at a glance that this is all part of the `Unsized` case:

```suggestion
        Unsized
          | i /= length (var ^. M.vfields) - 1 ->
            mirFail "encountered unsized field in non-last position"
          | otherwise ->
            case fldTy of
              M.TySlice innerTy -> do
                Some innerRepr <- tyToReprM innerTy
                pure $ UnsizedSliceField innerRepr
              M.TyStr -> do
                Some innerRepr <- tyToReprM (M.TyUint M.B8)
                pure $ UnsizedSliceField innerRepr
              _ -> pure UnsizedNonSliceField
```